src/D2L.CodeStyle.Analyzers/Language/OnlyVisibleToAnalyzer.Model.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+
+	public sealed partial class OnlyVisibleToAnalyzer {
+
+		internal readonly struct Model {
+
+			private const string OnlyVisibleToTypeAttributeMetadataName = "D2L.CodeStyle.Annotations.Contract.OnlyVisibleToTypeAttribute";
+
+			private readonly Compilation m_compilation;
+			private readonly INamedTypeSymbol m_onlyVisibleToTypeAttribute;
+
+			private readonly ConcurrentDictionary<ISymbol, ImmutableHashSet<INamedTypeSymbol>?> m_visibilityCache =
+				new ConcurrentDictionary<ISymbol, ImmutableHashSet<INamedTypeSymbol>?>( SymbolEqualityComparer.Default );
+
+			private Model(
+					Compilation compilation,
+					INamedTypeSymbol onlyVisibleToTypeAttribute
+				) {
+
+				m_compilation = compilation;
+				m_onlyVisibleToTypeAttribute = onlyVisibleToTypeAttribute;
+			}
+
+			public static Model? TryCreate( Compilation compilation ) {
+
+				INamedTypeSymbol? onlyVisibleToTypeAttribute = compilation.GetTypeByMetadataName( OnlyVisibleToTypeAttributeMetadataName );
+				if( onlyVisibleToTypeAttribute == null ) {
+					return null;
+				}
+
+				return new Model( compilation, onlyVisibleToTypeAttribute );
+			}
+
+			public bool IsVisibleTo( INamedTypeSymbol caller, ISymbol member ) {
+
+				ImmutableHashSet<INamedTypeSymbol>? restrictions = m_visibilityCache
+					.GetOrAdd( member, GetTypeVisibilityRestrictions );
+
+				if( restrictions == null ) {
+					return true;
+				}
+
+				if( restrictions.Contains( caller ) ) {
+					return true;
+				}
+
+				if( SymbolEqualityComparer.Default.Equals( caller, member.ContainingType ) ) {
+					return true;
+				}

```

Comment:
Just thought of it because it's similar, but I don't think we care about this situation (can work around with private helpers), right?

```csharp
public interface IFoo {
  [OnlyVisibleToType( "Some.Other.Type", "SomeAssembly" )]
  void RestrictedThing();

  void OtherThing();
}

public sealed class Foo : IFoo {
  void IFoo.RestrictedThing() {}

  void IFoo.OtherThing() {
    IFoo @this = this;

   @this.RestrictedThing(); // D2L0099
  }
}
```

Reply:
Yeah, you would need to use a private helper

Reply:
Having it behave sorta like `protected` where things are implicitly available to you when you implement/extend a base class would be how I'd expect it to work, I think, but I agree its probably not something we necessarily need to worry about now.

Reply:
Might be good to add that one as a test case if it isn't there (haven't gotten that far yet) even if only to document current behaviour

Reply:
Will do in a follow up